### PR TITLE
adding attach vs attest guidance

### LINKED ISSUE #371: Enforce Documentation: The right way to work with SBOM attestations in Enforce
We have had many customers struggle with sbom attestations in enforce. Multiple customers have asked very directly:
- what's the right way to do sbom attestations in enforce? I can't get my sboms ingested, i don't know if I should do cosign attest, attach, or upload. how do I verify? what about predicate types? 

I think we need a page in the Enforce docs on working with sboms. Guidance should be:
- the formats we understand and ingest (spdx, cyclone, other?)
- how to properly do the attestation with cosign (i think it's cosign attest)
- the fact that today, you have to have a policy defined to validate sbom attestations for enforce to ingest it (this should change at some point to be automatic)

### PR DESCRIPTION
**Pull request by SharpRake**
## Type of change

This is an overview of how Enforce treats SBOMs vs attestations, through the context of the `cosign attach` and `cosign attest` commands.

### What should this PR do?

resolves #371 

### How should this PR be tested?

This doesn't require a tech test, as the only commands given are examples. Please copyedit for style and grammar and all that, and I'm also kind of iffy on the title and overall structure. Plus, if anything seems wrong or missing please call it out!

### CODE REVIEW

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+draft: false
+images: []
+menu:
+  docs:
+    parent: "chainguard-enforce-kubernetes"
+weight: 600
+toc: true
+---
+
+In order to fully leverage the capabilities offered by Chainguard Enforce, it's important to establish a process for creating quality *Software Bills of Materials* (SBOMs). An SBOM is a formal record that contains the details and supply chain relationships (such as dependencies) of the components used in building given software.
```
Comment (ltagliaferri) [inline content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:16]: I think linking out to our SBOM resources would be helpful. Also "given software" sounds a little odd to me. If you want to keep it, i would rephrase it as "a given software artifact"

```suggestion
In order to fully leverage the capabilities offered by Chainguard Enforce, it's important to establish a process for creating quality *Software Bills of Materials* (SBOMs). An [SBOM](/open-source/sbom/) is a formal record that contains the details and supply chain relationships (such as dependencies) of the components used in building software.
```

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+[Cosign](/open-source/sigstore/cosign/an-introduction-to-cosign/)  — a part of the Sigstore project — supports software artifact signing, verification, and storage in an OCI (Open Container Initiative) registry. The `cosign` command line tool offers two subcommands that you can use to associate an SBOM with a container image and then upload them to a registry: `cosign attach` and `cosign attest`.
```
Comment (ltagliaferri) [inline content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:18]:
```suggestion
[Cosign](/open-source/sigstore/cosign/an-introduction-to-cosign/)  — a part of the Sigstore project — supports software artifact signing, verification, and storage in an [OCI (Open Container Initiative)](/open-source/oci/) registry. The `cosign` command line tool offers two subcommands that you can use to associate an SBOM with a container image and then upload them to a registry: `cosign attach` and `cosign attest`. 
```

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+[Cosign](/open-source/sigstore/cosign/an-introduction-to-cosign/)  — a part of the Sigstore project — supports software artifact signing, verification, and storage in an OCI (Open Container Initiative) registry. The `cosign` command line tool offers two subcommands that you can use to associate an SBOM with a container image and then upload them to a registry: `cosign attach` and `cosign attest`. 
+
+However, these commands don't work the same way, and Chainguard Enforce treats SBOMs that have been attached differently from those that have been attested. This guide outlines the differences between these two subcommands and provides guidance for when you might want to use one over the other.
+
+
+## SBOMs vs. Attestations
+
+An SBOM is essentially an electronic packing slip: it's a list of all the components that went into making a piece of given software. But unless you have some indication of when the software was produced, who produced it, and how it was produced, then you can't say with any certainty that the components listed in the SBOM are actually part of the software you're running.
```
Comment (ltagliaferri) [inline content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:25]:
```suggestion
An SBOM is essentially an electronic packing slip: it's a list of all the components that went into making a given piece of software. But unless you have some indication of when the software was produced, who produced it, and how it was produced, then you can't say with any certainty that the components listed in the SBOM are actually part of the software you're running.
```

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+An *attestation* allows the end users or consumers of a software artifact (in the context of this guide, an SBOM) to verify the quality of that artifact independently from the producer of the software. It also requires software producers to provide verifiable proof of the quality of their software.
+
+Put differently, an attestation is a written assurance of a software artifact's *provenance*, or the verifiable information about the artifact describing where, when, and how it was produced. You can think of an attestation as a proclamation that "software artifact X" was produced by "person Y" at "time Z".
+
+Both `cosign attest` and `cosign attach` associate an artifact with an image and upload it to a registry. However, `cosign attest` generates an in-toto attestation while `cosign attach` does not. `cosign attest` then attaches it to the provided image and uploads it to a registry as an OCI artifact with a `.att` extension.
```
Comment (ltagliaferri) [inline content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:31]:
```suggestion
Both `cosign attest` and `cosign attach` associate an artifact with an image and upload it to a registry. However, `cosign attest` generates an [in-toto attestation](https://in-toto.io/) while `cosign attach` does not. `cosign attest` then attaches it to the provided image and uploads it to a registry as an OCI artifact with a `.att` extension.
```

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+Conversely, Chainguard Enforce will ingest attached SBOMs out of the box, but cannot apply policies against them. With that said, if you can trust with absolute certainty that an unsigned SBOM is accurate (say, if you yourself created and attached it to the image) or if you don't care about policy enforcement, then attaching SBOMs may suit your needs. 
+
+Say you ran a `cosign attach` command to attach a `.sbom` artifact to an image and upload it to a registry. Then you ran `cosign sign` to sign the SBOM, generate a `.sig` artifact, and upload that to the registry. In this scenario, Chainguard Enforce would still see the SBOM as a `.sbom` artifact and not as an attestation (`.att`). Even though you've signed the SBOM, the signature is a separate artifact so Enforce cannot treat it as an attestation, and thus still won't be able to apply policies against it.
+
+The following table presents a high-level comparison of the `cosign attach` and `cosign attest` commands.
```
Comment (ltagliaferri) [inline content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:66]: I think the table is really useful, what do you think about adding this towards the top of the document in its own section?

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+There are many tools available today — both proprietary and open source — that allow you to generate SBOMs. However, thee tools do not generate SBOMs in the same way or at the same point in the development process. As with signed versus unsigned SBOMs, an organization may find SBOMs generated by one tool as more trustworthy than those from others. 
+
+For example, SBOMs generated from source code can be valuable. But ultimately, you have no way of knowing whether the image has been tampered with between the time the SBOM was generated and the time you actually run the image. 
+
+[apko](/open-source/apko/overview/) is a command-line tool that allows users to build container images using a declarative language based on YAML. When building a container image, apko will generate an SBOM outlining each of the apks it uses to build it. When combined with [melange](/open-source/melange/overview/), an apk builder tool that uses declarative pipelines to create apk packages, these tools can serve as a good starting point for a secure container image factory. Checkout [this blog post](https://www.chainguard.dev/unchained/secure-your-software-factory-with-melange-and-apko) to learn more.
```
Comment (ltagliaferri) [inline content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:84]: I think the title is more useful for describing the blog post

```suggestion
[apko](/open-source/apko/overview/) is a command-line tool that allows users to build container images using a declarative language based on YAML. When building a container image, apko will generate an SBOM outlining each of the apks it uses to build it. When combined with [melange](/open-source/melange/overview/), an apk builder tool that uses declarative pipelines to create apk packages, these tools can serve as a good starting point for a secure container image factory. Checkout ["Secure Your Software Factory with melange and apko"](https://www.chainguard.dev/unchained/secure-your-software-factory-with-melange-and-apko) to learn more.
```

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+|----------|----------|----------|
+| Produces an in-toto attestation (a "signed SBOM") | no  | yes  |
+| SBOM can be ingested immediately | yes  | no  |
+| SBOM requires policy for ingestion | no | yes  |
+| Policies can be applied to SBOM | no | yes |
+| Uploads OCI artifact to registry | yes | yes |
+| OCI artifact extension | `.sbom` | `.att` |
+
+
+## A Note on generating SBOMs
```
Comment (ltagliaferri) [inline content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:78]: I would lower case "Note" since the other headers only have title case on the first word.
```suggestion
## A note on generating SBOMs
```

**Review by ltagliaferri — APPROVED** [review]

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md (hunk 5 above)
Comment (SharpRake) [reply content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:66]: I'm for it!

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+images: []
+menu:
+  docs:
+    parent: "chainguard-enforce-kubernetes"
+weight: 600
+toc: true
+---
+
+In order to fully leverage the capabilities offered by Chainguard Enforce, it's important to establish a process for creating quality *Software Bills of Materials* (SBOMs). An [SBOM](/open-source/sbom/) is a formal record that contains the details and supply chain relationships (such as dependencies) of the components used in building software.
+
```
Comment (amdawson) [inline content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:17]: suggest putting a short sentence/para in here that says something like "CG Enforce can capture SBOM data and use it to help you find images and packages in your ecosystem, as well as apply security policies to your runtime environments. To get the best results from these features, SBOMs should be attached to the images in your registry as signed attestations." 

or something like that. just an overview of what this document is solving

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+## Comparison of attaching and attesting
+
+The following table presents a high-level comparison of the `cosign attach` and `cosign attest` commands.
+
+|   | `cosign attach` | `cosign attest` |
+|----------|----------|----------|
+| Produces an in-toto attestation (a "signed SBOM") | no  | yes  |
+| SBOM can be ingested immediately | yes  | no  |
```
Comment (amdawson) [inline content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:30]: @tcnghia or @mattmoor  can you review this table please for accuracy?

Comment (amdawson) [reply content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:30]: I think we ingest .sbom and .att now

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+An *attestation* allows the end users or consumers of a software artifact (in the context of this guide, an SBOM) to verify the quality of that artifact independently from the producer of the software. It also requires software producers to provide verifiable proof of the quality of their software.
+
+Put differently, an attestation is a written assurance of a software artifact's *provenance*, or the verifiable information about the artifact describing where, when, and how it was produced. You can think of an attestation as a proclamation that "software artifact X" was produced by "person Y" at "time Z".
+
+Both `cosign attest` and `cosign attach` associate an artifact with an image and upload it to a registry. However, `cosign attest` generates an [in-toto attestation](https://in-toto.io/) while `cosign attach` does not. `cosign attest` then attaches it to the provided image and uploads it to a registry as an OCI artifact with a `.att` extension.
```
Comment (amdawson) [inline content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:47]: this paragraph is excellent, could even go higher up in the doc.

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+`cosign attach`, on the other hand, only attaches an SBOM to an image and uploads it to a registry. In this case the SBOM isn't signed, meaning that Chainguard Enforce won't have any way to verify its authenticity.
+
+```sh
+cosign attach sbom --sbom image.sbom $IMAGE
+```
+
+This will upload the SBOM to the registry as an OCI artifact with a `.sbom` extension.
+
+Be aware that there is also the `cosign sign` command. After running `cosign attach` to attach an SBOM and upload it to a registry, you can then run `cosign sign` to sign the SBOM, and upload the signature to the registry as a separate OCI artifact, this time with the `.sig` extension.
```
Comment (amdawson) [inline content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:69]: might be helpful to briefly explain how enforce handles a .sig sbom...

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+
+## Should I attest or attach?
+
+Attestations are generally seen as being more trustworthy than unsigned SBOMs, because you can identify who signed them and when. However, Chainguard Enforce will not ingest an attestation unless there is an active policy covering it.
```
Comment (amdawson) [inline content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:74]: again, this may have changed. 

either way, perhaps me mention what "ingest" means, here or maybe above in the table section

**Review by amdawson — COMMENTED** [review]
great job, made some suggestions, and also would like someone on the enforce eng team to review the details of ingestion and policy coverage before publishing. suggest matt or nghia

File: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+| OCI artifact extension | `.sbom` | `.att` |
+
+The remaining sections of this guide elaborate further on the similarities and differences between these two commands.
+
+
+## SBOMs vs. Attestations
+
+An SBOM is essentially an electronic packing slip: it's a list of all the components that went into making a given piece of software. But unless you have some indication of when the software was produced, who produced it, and how it was produced, then you can't say with any certainty that the components listed in the SBOM are actually part of the software you're running.
+
+An *attestation* allows the end users or consumers of a software artifact (in the context of this guide, an SBOM) to verify the quality of that artifact independently from the producer of the software. It also requires software producers to provide verifiable proof of the quality of their software.
```
[the remark this answers is not in the document]
Comment (SharpRake) [reply content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/sboms-and-attestations.md:43]: That's way clearer, thanks Adam.

**Comment by SharpRake** [comment]
I'm still learning my way around GitHub and seem to have borked this PR. I've started a new PR with Adam's suggestions here: https://github.com/chainguard-dev/edu/pull/493
